File: src/fable-library-rust/src/BigInt.rs
#[cfg(feature = "bigint")]
pub mod BigInt_ {
    use crate::Decimal_::{decimal, truncate};
    use crate::NativeArray_::{array_from, Array};
    use crate::Native_::{compare, Lrc, MutCell, ToString, Vec};
    use crate::String_::{string, toString as toString_1};

    use num_bigint::*;
    use num_integer::*;
    use num_traits::*;

    #[repr(transparent)]
    #[derive(Clone, Debug, Default, Eq, Hash, PartialEq, PartialOrd, Ord)]
    pub struct bigint(Lrc<BigInt>);

    impl From<BigInt> for bigint {
        fn from(x: BigInt) -> bigint {
            bigint(Lrc::from(x))
        }
    }

    impl core::ops::Deref for bigint {
        type Target = Lrc<BigInt>;
        #[inline]
        fn deref(&self) -> &Self::Target {
            &self.0
        }
    }

    impl core::fmt::Display for bigint {
        fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
            write!(f, "{}", self.as_ref()) //TODO: improve
        }
    }

    macro_rules! un_op {
        ($op_trait:ident, $op_fn:ident, $op:tt) => {
            impl core::ops::$op_trait for bigint {
                type Output = Self;
                #[inline]
                fn $op_fn(self) -> Self::Output {
                    ($op self.as_ref()).into()
                }
            }
        };
    }

    macro_rules! bin_op {
        ($op_trait:ident, $op_fn:ident, $op:tt) => {
            impl core::ops::$op_trait for bigint {
                type Output = Self;
                #[inline]
                fn $op_fn(self, rhs: Self) -> Self::Output {
                    (self.as_ref() $op rhs.as_ref()).into()
                }
            }
        };
    }

    macro_rules! shift_op {
        ($op_trait:ident, $op_fn:ident, $op:tt) => {
            impl core::ops::$op_trait<i32> for bigint {
                type Output = Self;
                #[inline]
                fn $op_fn(self, rhs: i32) -> Self::Output {
                    (self.as_ref() $op rhs).into()
                }
            }
        };
    }

    un_op!(Neg, neg, -);
    un_op!(Not, not, !);

    bin_op!(Add, add, +);
    bin_op!(Sub, sub, -);
    bin_op!(Mul, mul, *);
    bin_op!(Div, div, /);
    bin_op!(Rem, rem, %);

    bin_op!(BitAnd, bitand, &);
    bin_op!(BitOr, bitor, |);
    bin_op!(BitXor, bitxor, ^);

    shift_op!(Shl, shl, <<);
    shift_op!(Shr, shr, >>);

    pub fn zero() -> bigint { BigInt::zero().into() }
    pub fn one() -> bigint { BigInt::one().into() }
    pub fn minusOne() -> bigint { (-BigInt::one()).into() }

    // pub fn getHashCode(x: bigint) = x.GetHashCode()
    pub fn equals(x: bigint, y: bigint) -> bool { x.eq(&y) }
    pub fn compareTo(x: bigint, y: bigint) -> i32 { compare(&x, &y) }

    pub fn abs(x: bigint) -> bigint { x.abs().into() }
    pub fn sign(x: bigint) -> i32 { x.signum().to_i32().unwrap() }

    pub fn max(x: bigint, y: bigint) -> bigint { x.max(y).into() }
    pub fn min(x: bigint, y: bigint) -> bigint { x.min(y).into() }

    pub fn maxMagnitude(x: bigint, y: bigint) -> bigint {
        if abs(x.clone()) > abs(y.clone()) { x } else { y }
    }

    pub fn minMagnitude(x: bigint, y: bigint) -> bigint {
        if abs(x.clone()) < abs(y.clone()) { x } else { y }
    }

    pub fn clamp(x: bigint, min: bigint, max: bigint) -> bigint {
        x.clamp(min, max).into()
    }

    pub fn add(x: bigint, y: bigint) -> bigint { x + y }
    pub fn subtract(x: bigint, y: bigint) -> bigint { x - y }
    pub fn multiply(x: bigint, y: bigint) -> bigint { x * y }
    pub fn divide(x: bigint, y: bigint) -> bigint { x / y }
    pub fn remainder(x: bigint, y: bigint) -> bigint { x % y }
    pub fn negate(x: bigint) -> bigint { -x }

    pub fn isZero(x: bigint) -> bool { x.is_zero() }
    pub fn isOne(x: bigint) -> bool { x.is_one() }
    pub fn isEven(x: bigint) -> bool { x.is_even() }
    pub fn isPowerOfTwo(x: bigint) -> bool { isPow2(x) }

    pub fn isNegative(x: bigint) -> bool { x.is_negative() }
    pub fn isPositive(x: bigint) -> bool { x.is_positive() }
    pub fn isEvenInteger(x: bigint) -> bool { x.is_even() }
    pub fn isOddInteger(x: bigint) -> bool { x.is_odd() }
    pub fn isPow2(x: bigint) -> bool { (x.clone() & (x - one())).is_zero() }

    pub fn fromZero() -> bigint { BigInt::zero().into() }
    pub fn fromOne() -> bigint { BigInt::one().into() }

    pub fn fromInt8(n: i8) -> bigint { BigInt::from_i8(n).unwrap().into() }
    pub fn fromUInt8(n: u8) -> bigint { BigInt::from_u8(n).unwrap().into() }
    pub fn fromInt16(n: i16) -> bigint { BigInt::from_i16(n).unwrap().into() }
    pub fn fromUInt16(n: u16) -> bigint { BigInt::from_u16(n).unwrap().into() }
    pub fn fromInt32(n: i32) -> bigint { BigInt::from_i32(n).unwrap().into() }
    pub fn fromUInt32(n: u32) -> bigint { BigInt::from_u32(n).unwrap().into() }
    pub fn fromInt64(n: i64) -> bigint { BigInt::from_i64(n).unwrap().into() }
    pub fn fromUInt64(n: u64) -> bigint { BigInt::from_u64(n).unwrap().into() }
    pub fn fromNativeInt(n: isize) -> bigint { BigInt::from_isize(n).unwrap().into() }
    pub fn fromUNativeInt(n: usize) -> bigint { BigInt::from_usize(n).unwrap().into() }

    pub fn fromFloat32(n: f32) -> bigint { BigInt::from_f32(n).unwrap().into() }
    pub fn fromFloat64(n: f64) -> bigint { BigInt::from_f64(n).unwrap().into() }

    pub fn fromDecimal(d: decimal) -> bigint {
        BigInt::from_str_radix(truncate(d).to_string().as_str(), 10).unwrap().into()
    }

    pub fn fromBigInt(x: bigint) -> bigint { x }

    pub fn fromBoolean(b: bool) -> bigint {
        BigInt::from_u32(b as u32).unwrap().into()
    }

    pub fn fromChar(c: char) -> bigint {
        BigInt::from_u32(c as u32).unwrap().into()
    }

    pub fn fromString(s: string) -> bigint {
        BigInt::from_str_radix(s.trim(), 10).unwrap().into()
    }

    pub fn fromByteArray(bytes: Array<u8>) -> bigint {
        BigInt::from_signed_bytes_le(bytes.as_ref()).into()
    }

    pub fn toByteArray(x: bigint) -> Array<u8> {
        array_from(x.to_signed_bytes_le())
    }

    pub fn toInt8(x: bigint) -> i8 { x.to_i8().unwrap() }
    pub fn toUInt8(x: bigint) -> u8 { x.to_u8().unwrap() }
    pub fn toInt16(x: bigint) -> i16 { x.to_i16().unwrap() }
    pub fn toUInt16(x: bigint) -> u16 { x.to_u16().unwrap() }
    pub fn toInt32(x: bigint) -> i32 { x.to_i32().unwrap() }
    pub fn toUInt32(x: bigint) -> u32 { x.to_u32().unwrap() }
    pub fn toInt64(x: bigint) -> i64 { x.to_i64().unwrap() }
    pub fn toUInt64(x: bigint) -> u64 { x.to_u64().unwrap() }
    pub fn toNativeInt(x: bigint) -> isize { x.to_isize().unwrap() }
    pub fn toUNativeInt(x: bigint) -> usize { x.to_usize().unwrap() }
    pub fn toFloat32(x: bigint) -> f32 { x.to_f32().unwrap() }
    pub fn toFloat64(x: bigint) -> f64 { x.to_f64().unwrap() }

    pub fn toDecimal(x: bigint) -> decimal {
        decimal::from_str_radix(x.to_string().as_str(), 10).unwrap()
    }

    pub fn toBigInt(x: bigint) -> bigint { x }

    pub fn toBoolean(x: bigint) -> bool { !x.is_zero() }

    pub fn toChar(x: bigint) -> char {
        core::char::from_u32(x.to_u32().unwrap()).unwrap()
    }

    pub fn toString(x: bigint) -> string {
        toString_1(&x)
    }

    pub fn tryParse(s: string, res: &MutCell<bigint>) -> bool {
        match BigInt::from_str_radix(s.trim(), 10) {
            Ok(d) => { res.set(d.into()); true },
            Err(e) => false,
        }
    }

    pub fn parse(s: string) -> bigint {
        match BigInt::from_str_radix(s.trim(), 10) {
            Ok(d) => d.into(),
            Err(e) => panic!("The input string {} was not in a correct format.", s),
        }
    }

    pub fn pow(x: bigint, n: i32) -> bigint {
        if n < 0 {
            panic!("The exponent must be greater than or equal to zero.")
        } else {
            x.0.as_ref().pow(n as u32).into()
        }
    }

    pub fn modPow(x: bigint, e: bigint, m: bigint) -> bigint {
        x.modpow(&e, &m).into()
    }

    pub fn divRem(x: bigint, y: bigint) -> (bigint, bigint) {
        let (div, rem) = x.div_rem(&y);
        (div.into(), rem.into())
    }

    pub fn divRemOut(x: bigint, y: bigint, remainder: &MutCell<bigint>) -> bigint {
        let (div, rem) = x.div_rem(&y);
        remainder.set(rem.into());
        div.into()
    }

    pub fn greatestCommonDivisor(x: bigint, y: bigint) -> bigint {
        x.gcd(&y).into()
    }

    pub fn getBitLength(x: bigint) -> i64 {
        x.bits() as i64
    }

    pub fn log2(x: bigint) -> f64 {
        if x.is_negative() {
            f64::NAN
        } else {
            let bits = x.bits();
            if bits < 1024 {
                x.to_f64().unwrap().log2()
            } else {
                (bits - 1) as f64
            }
        }
    }

    pub fn log10(x: bigint) -> f64 {
        log2(x) * 2_f64.log10()
    }

    pub fn ln(x: bigint) -> f64 {
        log2(x) * 2_f64.ln()
    }

    pub fn log(x: bigint, base: f64) -> f64 {
        log2(x) / base.log2()
    }

    pub fn ilog2(x: bigint) -> bigint {
        fromFloat64(log2(x))
    }

    // pub fn copySign
    // pub fn createChecked
    // pub fn createSaturating
    // pub fn createTruncating
    // pub fn getByteCount
    // pub fn leadingZeroCount
    // pub fn popCount
    // pub fn rotateLeft
    // pub fn rotateRight
    // pub fn trailingZeroCount
    // pub fn tryFormat
    // pub fn tryWriteBytes
}

File: src/fable-compiler-js/fable-library/BigInt.js
import { fromParts, truncate } from "./Decimal.js";
import { bigintHash } from "./Util.js";
const isBigEndian = false;
BigInt.prototype.toJSON = function () {
    return `${this.toString()}`;
};
const zero = 0n;
const one = 1n;
const two = 2n;
const minusOne = -1n;
export function isBigInt(x) {
    return typeof x === "bigint";
}
export function hash(x) {
    return bigintHash(x);
}
export function equals(x, y) {
    return x === y;
}
export function compare(x, y) {
    return x < y ? -1 : x > y ? 1 : 0;
}
export function abs(x) { return x < zero ? -x : x; }
export function sign(x) { return x < zero ? -1 : x > zero ? 1 : 0; }
export function max(x, y) { return x > y ? x : y; }
export function min(x, y) { return x < y ? x : y; }
export function maxMagnitude(x, y) { return abs(x) > abs(y) ? x : y; }
export function minMagnitude(x, y) { return abs(x) < abs(y) ? x : y; }
export function clamp(x, min, max) {
    return x < min ? min : x > max ? max : x;
}
export function add(x, y) { return x + y; }
export function subtract(x, y) { return x - y; }
export function multiply(x, y) { return x * y; }
export function divide(x, y) { return x / y; }
export function remainder(x, y) { return x % y; }
export function negate(x) { return -x; }
export function op_UnaryNegation(x) { return -x; }
export function op_LogicalNot(x) { return ~x; }
export function op_UnaryPlus(x) { return x; }
export function op_Addition(x, y) { return x + y; }
export function op_Subtraction(x, y) { return x - y; }
export function op_Multiply(x, y) { return x * y; }
export function op_Division(x, y) { return x / y; }
export function op_Modulus(x, y) { return x % y; }
export function op_RightShift(x, n) { return x >> BigInt(n); }
export function op_LeftShift(x, n) { return x << BigInt(n); }
export function op_BitwiseAnd(x, y) { return x & y; }
export function op_BitwiseOr(x, y) { return x | y; }
export function op_ExclusiveOr(x, y) { return x ^ y; }
export function op_LessThan(x, y) { return x < y; }
export function op_LessThanOrEqual(x, y) { return x <= y; }
export function op_GreaterThan(x, y) { return x > y; }
export function op_GreaterThanOrEqual(x, y) { return x >= y; }
export function op_Equality(x, y) { return x === y; }
export function op_Inequality(x, y) { return x !== y; }
export function get_Zero() { return zero; }
export function get_One() { return one; }
export function get_MinusOne() { return minusOne; }
export function get_IsZero(x) { return x === zero; }
export function get_IsOne(x) { return x === one; }
export function get_IsEven(x) { return isEvenInteger(x); }
export function get_IsPowerOfTwo(x) { return isPow2(x); }
export function get_Sign(x) { return sign(x); }
export function isNegative(x) { return x < zero; }
export function isPositive(x) { return x > zero; }
export function isEvenInteger(x) { return (x % two) === zero; }
export function isOddInteger(x) { return (x % two) !== zero; }
export function isPow2(x) { return (x & (x - one)) === zero; }
export function fromZero() { return zero; }
export function fromOne() { return one; }
export function fromInt8(n) { return BigInt(n); }
export function fromUInt8(n) { return BigInt(n); }
export function fromInt16(n) { return BigInt(n); }
export function fromUInt16(n) { return BigInt(n); }
export function fromInt32(n) { return BigInt(n); }
export function fromUInt32(n) { return BigInt(n); }
export function fromInt64(n) { return n; }
export function fromUInt64(n) { return n; }
export function fromInt128(n) { return n; }
export function fromUInt128(n) { return n; }
export function fromNativeInt(n) { return n; }
export function fromUNativeInt(n) { return n; }
export function fromFloat16(n) { return BigInt(Math.trunc(n)); }
export function fromFloat32(n) { return BigInt(Math.trunc(n)); }
export function fromFloat64(n) { return BigInt(Math.trunc(n)); }
export function fromDecimal(d) { return BigInt(truncate(d).toString()); }
export function fromBigInt(x) { return x; }
export function fromBoolean(b) { return BigInt(b); }
export function fromChar(c) { return BigInt(c.charCodeAt(0)); }
export function fromString(s) { return BigInt(s); }
export function fromByteArray(bytes) {
    return fromSignedBytes(bytes, isBigEndian);
}
export function toByteArray(value) {
    return toSignedBytes(value, isBigEndian);
}
export function toInt8(x) { return Number(BigInt.asIntN(8, x)); }
export function toUInt8(x) { return Number(BigInt.asUintN(8, x)); }
export function toInt16(x) { return Number(BigInt.asIntN(16, x)); }
export function toUInt16(x) { return Number(BigInt.asUintN(16, x)); }
export function toInt32(x) { return Number(BigInt.asIntN(32, x)); }
export function toUInt32(x) { return Number(BigInt.asUintN(32, x)); }
export function toInt64(x) { return BigInt.asIntN(64, x); }
export function toUInt64(x) { return BigInt.asUintN(64, x); }
export function toInt128(x) { return BigInt.asIntN(128, x); }
export function toUInt128(x) { return BigInt.asUintN(128, x); }
export function toNativeInt(x) { return BigInt.asIntN(64, x); }
export function toUNativeInt(x) { return BigInt.asUintN(64, x); }
export function toFloat16(x) { return Number(x); }
export function toFloat32(x) { return Number(x); }
export function toFloat64(x) { return Number(x); }
export function toDecimal(x) {
    const low = Number(BigInt.asUintN(32, x));
    const mid = Number(BigInt.asUintN(32, x >> 32n));
    const high = Number(BigInt.asUintN(32, x >> 64n));
    const isNegative = x < zero;
    const scale = 0;
    return fromParts(low, mid, high, isNegative, scale);
}
export function toBigInt(x) { return x; }
export function toBoolean(x) { return x !== zero; }
export function toChar(x) {
    return String.fromCharCode(toUInt16(x));
}
export function toString(x) { return x.toString(); }
export function tryParse(s, res) {
    try {
        res.contents = BigInt(s);
        return true;
    }
    catch (err) {
        return false;
    }
}
export function parse(s) {
    return BigInt(s);
}
export function pow(x, n) {
    return x ** BigInt(n);
}
export function modPow(x, e, m) {
    return (x ** e) % m;
}
export function divRem(x, y, out) {
    const div = x / y;
    const rem = x % y;
    if (out === void 0) {
        return [div, rem];
    }
    else {
        out.contents = rem;
        return div;
    }
}
export function greatestCommonDivisor(x, y) {
    while (y > zero) {
        const q = x / y;
        const r = x - q * y;
        x = y;
        y = r;
    }
    return x;
}
export function getBitLength(x) {
    return fromFloat64(x === zero ? 1 : log2(abs(x)) + 1);
}
export function log2(x) {
    const n = Number(x);
    if (Number.isFinite(n))
        return Math.log2(n); // fast path
    if (x < zero)
        return Number.NaN;
    let shift = one;
    while (x >= (one << shift)) {
        shift = shift << one;
    }
    let log = zero;
    while (shift > one) {
        shift = shift >> one;
        if (x >= (one << shift)) {
            log = log + shift;
            x = x >> shift;
        }
    }
    return Number(log);
}
export function log10(x) {
    return log2(x) * Math.log10(2);
}
export function ln(x) {
    return log2(x) * Math.log(2);
}
export function log(x, base) {
    return log2(x) / Math.log2(base);
}
export function ilog2(x) {
    return BigInt(log2(x));
}
// export function copySign
// export function createChecked
// export function createSaturating
// export function createTruncating
// export function getByteCount
// export function leadingZeroCount
// export function popCount
// export function rotateLeft
// export function rotateRight
// export function trailingZeroCount
// export function tryFormat
// export function tryWriteBytes
// -------------------------------------------------
// Binary serialization
// -------------------------------------------------
const hexCodes = new Uint8Array([48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 97, 98, 99, 100, 101, 102]);
function fromHexCode(code) {
    if (48 <= code && code <= 57)
        return code - 48;
    if (97 <= code && code <= 102)
        return code - 97 + 10;
    if (65 <= code && code <= 70)
        return code - 65 + 10;
    throw Error(`Invalid hex code: ${code}`);
}
function toSignedBytes(x, isBigEndian) {
    const isNeg = x < 0n;
    if (isNeg) {
        const len = log2(-x);
        const bits = len + (8 - len % 8);
        const pow2 = (1n << BigInt(bits));
        x = x + pow2; // two's complement
    }
    const hex = x.toString(16);
    const len = hex.length;
    const odd = len % 2;
    const first = hex.charCodeAt(0);
    const isLow = 48 <= first && first <= 55; // 0..7
    const start = (isNeg && isLow) || (!isNeg && !isLow) ? 1 : 0;
    const bytes = new Uint8Array(start + (len + odd) / 2);
    const inc = isBigEndian ? 1 : -1;
    let pos = isBigEndian ? 0 : bytes.length - 1;
    if (start > 0) {
        bytes[pos] = isNeg ? 255 : 0;
        pos += inc;
    }
    if (odd > 0) {
        bytes[pos] = fromHexCode(first);
        pos += inc;
    }
    for (let i = odd; i < len; i += 2, pos += inc) {
        const a = fromHexCode(hex.charCodeAt(i));
        const b = fromHexCode(hex.charCodeAt(i + 1));
        bytes[pos] = (a << 4) | b;
    }
    return bytes;
}
function fromSignedBytes(bytes, isBigEndian) {
    if (bytes == null) {
        throw new Error("bytes is null");
    }
    const len = bytes.length;
    const first = isBigEndian ? 0 : len - 1;
    const isNeg = bytes[first] > 127;
    const codes = new Uint16Array(len * 2 + 2);
    codes[0] = 48; // 0
    codes[1] = 120; // x
    const inc = isBigEndian ? 1 : -1;
    let pos = isBigEndian ? 0 : len - 1;
    for (let i = 0; i < bytes.length; i++, pos += inc) {
        const byte = bytes[pos];
        codes[2 * i + 2] = hexCodes[byte >> 4];
        codes[2 * i + 3] = hexCodes[byte & 15];
    }
    const str = String.fromCharCode.apply(null, codes);
    let x = BigInt(str);
    if (isNeg) {
        const bits = len * 8;
        const pow2 = (1n << BigInt(bits));
        x = x - pow2; // two's complement
    }
    return x;
}
